File: audits/clusterprofiler/verify/gsea_vs_reference.R
## GSEA core on one synthetic ranked list with known truth, run through every engine that
## computes published GSEA numbers for clusterProfiler users:
##   fgsea::fgseaMultilevel / fgseaSimple           (fgsea master; also the engine of clusterProfiler <= 4.18)
##   enrichit::gsea  method = multilevel | sample | permute   (engine of clusterProfiler >= 4.19.3)
##   DOSE:::GSEA_internal by = "fgsea" | "DOSE"    (legacy DOSE 3.30.0, if lib-legacy is on R_LIBS_USER)
##   clusterProfiler::GSEA(TERM2GENE)                (devel HEAD, if installed)
## Everything is dumped as TSV; gsea_vs_reference.py recomputes ES, leading edge, rank,
## tags/list/signal and set sizes from Subramanian et al. 2005 and compares.
## Usage: R_LIBS_USER=<lib-legacy>:<lib> Rscript gsea_vs_reference.R <outdir>
args <- commandArgs(TRUE); outdir <- if (length(args)) args[1] else tempdir()
dir.create(outdir, showWarnings = FALSE, recursive = TRUE)
suppressPackageStartupMessages({library(fgsea); library(enrichit); library(jsonlite)})
cat("fgsea", as.character(packageVersion("fgsea")), " enrichit", as.character(packageVersion("enrichit")), "\n")
set.seed(1)
N <- 2000
stats <- rnorm(N)
names(stats) <- sprintf("g%04d", 1:N)
## planted signal: genes 1..60 shifted up, 61..100 shifted down
stats[1:60] <- stats[1:60] + 2.5; stats[61:100] <- stats[61:100] - 2.5
stats <- sort(stats, decreasing = TRUE)
genes <- names(stats)
sets <- list()
set.seed(2)
for (i in 1:30) sets[[sprintf("R%02d", i)]] <- sample(genes, sample(10:300, 1))          # random
sets$UP1   <- c(sprintf("g%04d", 1:40),  sample(genes, 30))                               # enriched at top
sets$DOWN1 <- c(sprintf("g%04d", 61:95), sample(genes, 25))                               # enriched at bottom
sets$ABSENT <- c(sample(genes, 12), sprintf("zz%02d", 1:8))                               # 8 genes not in the list
sets$DUP    <- rep(sample(genes, 15), 2)                                                  # duplicated members
sets$SMALL  <- sample(genes, 6)                                                           # below minGSSize 10
sets$BIG    <- sample(genes, 600)                                                         # above maxGSSize 500
sets <- lapply(sets, function(s) unique(s))  # fgsea documents that duplicates are removed; keep DUP explicit below
sets$DUP <- rep(sets$DUP, 2)
write_json(list(stats = as.list(stats), sets = sets), file.path(outdir, "design.json"), digits = NA)

dump <- function(df, name) write.table(df, file.path(outdir, paste0(name, ".tsv")), sep = "\t", quote = FALSE, row.names = FALSE)
as_chr <- function(l) vapply(l, paste, character(1), collapse = "/")

## ---- fgsea ---------------------------------------------------------------------------------
set.seed(11)
fm <- fgseaMultilevel(sets, stats, minSize = 10, maxSize = 500, eps = 0)
fm$leadingEdge <- as_chr(fm$leadingEdge); dump(as.data.frame(fm), "fgsea_multilevel")
set.seed(12)
fs <- fgseaSimple(sets, stats, nperm = 20000, minSize = 10, maxSize = 500)
fs$leadingEdge <- as_chr(fs$leadingEdge); dump(as.data.frame(fs), "fgsea_simple")
for (gp in c(0, 2)) {
    set.seed(13)
    f <- fgseaSimple(sets, stats, nperm = 2000, minSize = 10, maxSize = 500, gseaParam = gp)
    f$leadingEdge <- as_chr(f$leadingEdge); dump(as.data.frame(f), sprintf("fgsea_simple_p%d", gp))
}
set.seed(14); fpos <- fgseaMultilevel(sets, stats, minSize = 10, maxSize = 500, eps = 0, scoreType = "pos")
fpos$leadingEdge <- as_chr(fpos$leadingEdge); dump(as.data.frame(fpos), "fgsea_multilevel_pos")
cat(sprintf("fgsea: multilevel rows=%d simple rows=%d\n", nrow(fm), nrow(fs)))

## ---- enrichit ------------------------------------------------------------------------------
em <- suppressWarnings(enrichit::gsea(stats, sets, minGSSize = 10, maxGSSize = 500, method = "multilevel", seed = 21, eps = 0, verbose = FALSE))
dump(em, "enrichit_multilevel")
es <- suppressWarnings(enrichit::gsea(stats, sets, minGSSize = 10, maxGSSize = 500, method = "sample", nPerm = 20000, seed = 22, verbose = FALSE))
dump(es, "enrichit_sample")
ep <- suppressWarnings(enrichit::gsea(stats, sets, minGSSize = 10, maxGSSize = 500, method = "permute", nPerm = 2000, seed = 23, verbose = FALSE))
dump(ep, "enrichit_permute")
for (ex in c(0, 2)) {
    e <- suppressWarnings(enrichit::gsea(stats, sets, minGSSize = 10, maxGSSize = 500, method = "sample", nPerm = 2000, exponent = ex, seed = 24, verbose = FALSE))
    dump(e, sprintf("enrichit_sample_p%d", ex))
    e <- suppressWarnings(enrichit::gsea(stats, sets, minGSSize = 10, maxGSSize = 500, method = "multilevel", exponent = ex, seed = 24, eps = 0, verbose = FALSE))
    dump(e, sprintf("enrichit_multilevel_p%d", ex))
}
epos <- suppressWarnings(enrichit::gsea(stats, sets, minGSSize = 10, maxGSSize = 500, method = "multilevel", seed = 25, eps = 0, scoreType = "pos", verbose = FALSE))
dump(epos, "enrichit_multilevel_pos")
cat(sprintf("enrichit: multilevel rows=%d sample rows=%d permute rows=%d\n", nrow(em), nrow(es), nrow(ep)))

## seeds: FALSE draws from R's RNG; numeric seed fixes; set.seed() before the call also fixes
a1 <- suppressWarnings(enrichit::gsea(stats, sets["UP1"], method = "multilevel", eps = 0, verbose = FALSE))
a2 <- suppressWarnings(enrichit::gsea(stats, sets["UP1"], method = "multilevel", eps = 0, verbose = FALSE))
b1 <- suppressWarnings(enrichit::gsea(stats, sets["UP1"], method = "multilevel", eps = 0, seed = 7, verbose = FALSE))
b2 <- suppressWarnings(enrichit::gsea(stats, sets["UP1"], method = "multilevel", eps = 0, seed = 7, verbose = FALSE))
set.seed(3); c1 <- suppressWarnings(enrichit::gsea(stats, sets["UP1"], method = "multilevel", eps = 0, verbose = FALSE))
set.seed(3); c2 <- suppressWarnings(enrichit::gsea(stats, sets["UP1"], method = "multilevel", eps = 0, verbose = FALSE))
cat(sprintf("enrichit seed: seed=FALSE twice identical p: %s (%.3e vs %.3e); seed=7 twice: %s; set.seed(3)+FALSE twice: %s\n",
            identical(a1$pvalue, a2$pvalue), a1$pvalue, a2$pvalue, identical(b1$pvalue, b2$pvalue), identical(c1$pvalue, c2$pvalue)))

## ties: two ways of ordering the same tied block must give the ES of the input order
st <- stats; st[c("g0005", "g0006", "g0007")] <- st["g0005"]   # a 3-way tie near the top
tie_set <- list(T = c("g0005", "g0006", "g0007", sample(genes[200:2000], 20)))
w <- NULL
r1 <- withCallingHandlers(fgseaSimple(tie_set, st, nperm = 10, minSize = 1), warning = function(x) { w <<- c(w, conditionMessage(x)); invokeRestart("muffleWarning") })
r2 <- withCallingHandlers(enrichit::gsea(st, tie_set, minGSSize = 1, method = "sample", nPerm = 10, verbose = FALSE), warning = function(x) { w <<- c(w, conditionMessage(x)); invokeRestart("muffleWarning") })
cat(sprintf("ties: fgsea ES=%.6f enrichit ES=%.6f; warnings issued: %d (%s)\n", r1$ES, r2$enrichmentScore, length(w), paste(substr(unique(w), 1, 40), collapse = " | ")))
write_json(list(stats = as.list(st), set = tie_set$T, fgsea_ES = r1$ES, enrichit_ES = r2$enrichmentScore), file.path(outdir, "ties.json"), digits = NA)

## ---- legacy DOSE 3.30.0 (fgsea backend and its own permutation backend) --------------------
if (requireNamespace("DOSE", quietly = TRUE) && packageVersion("DOSE") < "4.0.0") {
    cat("DOSE", as.character(packageVersion("DOSE")), "\n")
    gs <- gson::gson(gsid2gene = data.frame(gsid = rep(names(sets), lengths(sets)), gene = unlist(sets)),
                     gsid2name = data.frame(gsid = names(sets), name = names(sets)),
                     species = "s", gsname = "s", version = "0", accessed_date = "2026-09-08", keytype = "k")
    GI <- get("GSEA_internal", envir = asNamespace("DOSE"))
    set.seed(31)
    d1 <- GI(geneList = stats, exponent = 1, minGSSize = 10, maxGSSize = 500, eps = 0, pvalueCutoff = 1,
             pAdjustMethod = "BH", verbose = FALSE, USER_DATA = gs, by = "fgsea")
    dump(d1@result, "dose_fgsea")
    set.seed(32)
    d2 <- GI(geneList = stats, exponent = 1, minGSSize = 10, maxGSSize = 500, eps = 0, pvalueCutoff = 1,
             pAdjustMethod = "BH", verbose = FALSE, USER_DATA = gs, by = "DOSE", nPerm = 2000)
    dump(d2@result, "dose_perm")
    cat(sprintf("DOSE: by=fgsea rows=%d by=DOSE rows=%d\n", nrow(d1@result), nrow(d2@result)))
}
## ---- clusterProfiler devel wrapper --------------------------------------------------------
if (requireNamespace("clusterProfiler", quietly = TRUE) && packageVersion("clusterProfiler") >= "4.19.3") {
    suppressPackageStartupMessages(library(clusterProfiler))
    cat("clusterProfiler", as.character(packageVersion("clusterProfiler")), "\n")
    t2g <- data.frame(term = rep(names(sets), lengths(sets)), gene = unlist(sets))
    cp <- suppressWarnings(GSEA(stats, TERM2GENE = t2g, pvalueCutoff = 1, eps = 0, seed = 21, verbose = FALSE))
    dump(cp@result, "clusterProfiler_GSEA")
    cat(sprintf("clusterProfiler::GSEA rows=%d; identical to enrichit multilevel (same seed): ES %s, pvalue %s\n",
                nrow(cp@result), isTRUE(all.equal(cp@result[em$ID, "enrichmentScore"], em$enrichmentScore)),
                isTRUE(all.equal(cp@result[em$ID, "pvalue"], em$pvalue))))
}
cat("done\n")

File: audits/clusterprofiler/verify/gsea_vs_reference.py
"""Compare the GSEA tables written by gsea_vs_reference.R with the numpy reference (ref_gsea.py).
Usage: python gsea_vs_reference.py <outdir>"""
import json, sys, os
import numpy as np, pandas as pd
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from ref_gsea import gsea_es

out = sys.argv[1]
d = json.load(open(os.path.join(out, "design.json")))
stats = pd.Series({k: v[0] if isinstance(v, list) else v for k, v in d["stats"].items()})
stats = stats.sort_values(ascending=False)          # the R script sorted before dumping; keep names
genes = np.array(stats.index); sv = stats.values
sets = {k: list(v) for k, v in d["sets"].items()}


def ref(setname, p=1.0):
    members = set(sets[setname]) & set(genes)
    hits = np.isin(genes, list(members))
    r = gsea_es(sv, hits, p)
    r["size"] = int(hits.sum())
    r["le_genes"] = set(genes[r["leading_edge_idx"]])
    return r


def load(name):
    f = os.path.join(out, name + ".tsv")
    return pd.read_csv(f, sep="\t") if os.path.exists(f) else None


def report(name, idcol, escol, sizecol=None, lecol=None, rankcol=None, lestr=None, p=1.0, note=""):
    df = load(name)
    if df is None:
        print(f"\n== {name}: not produced"); return None
    df = df.set_index(idcol)
    print(f"\n== {name}{note}: {len(df)} sets tested; SMALL tested: {'SMALL' in df.index}; BIG tested: {'BIG' in df.index}")
    dES = []; dsize = []; le_ok = []; rank_ok = []; tls_ok = []
    for sid in df.index:
        r = ref(sid, p)
        dES.append(abs(df.loc[sid, escol] - r["ES"]))
        if sizecol: dsize.append(int(df.loc[sid, sizecol]) - r["size"])
        if lecol and isinstance(df.loc[sid, lecol], str):
            le_ok.append(set(df.loc[sid, lecol].split("/")) == r["le_genes"])
        if rankcol: rank_ok.append(int(df.loc[sid, rankcol]) == r["rank"])
        if lestr:
            s = df.loc[sid, lestr]
            want = f"tags={round(r['tags']*100)}%, list={round(r['list']*100)}%, signal={round(r['signal']*100)}%"
            tls_ok.append(s == want)
    print(f"   max |ES - reference| = {max(dES):.2e}")
    if sizecol: print(f"   size - reference size: {sorted(set(dsize))}  (ABSENT size = {int(df.loc['ABSENT', sizecol]) if 'ABSENT' in df.index else 'n/a'}, DUP size = {int(df.loc['DUP', sizecol]) if 'DUP' in df.index else 'n/a'})")
    if lecol: print(f"   leading edge == reference: {sum(le_ok)}/{len(le_ok)}")
    if rankcol: print(f"   rank == reference: {sum(rank_ok)}/{len(rank_ok)}")
    if lestr: print(f"   'tags/list/signal' string == reference: {sum(tls_ok)}/{len(tls_ok)}")
    return df


fm = report("fgsea_multilevel", "pathway", "ES", "size", "leadingEdge")
fs = report("fgsea_simple", "pathway", "ES", "size", "leadingEdge")
report("fgsea_simple_p0", "pathway", "ES", "size", "leadingEdge", p=0, note=" (gseaParam=0)")
report("fgsea_simple_p2", "pathway", "ES", "size", "leadingEdge", p=2, note=" (gseaParam=2)")
em = report("enrichit_multilevel", "ID", "enrichmentScore", "setSize", "core_enrichment", "rank", "leading_edge")
es = report("enrichit_sample", "ID", "enrichmentScore", "setSize", "core_enrichment", "rank", "leading_edge")
ep = report("enrichit_permute", "ID", "enrichmentScore", "setSize", "core_enrichment", "rank", "leading_edge")
report("enrichit_sample_p0", "ID", "enrichmentScore", "setSize", "core_enrichment", "rank", "leading_edge", p=0, note=" (exponent=0)")
report("enrichit_sample_p2", "ID", "enrichmentScore", "setSize", "core_enrichment", "rank", "leading_edge", p=2, note=" (exponent=2)")
report("enrichit_multilevel_p0", "ID", "enrichmentScore", "setSize", "core_enrichment", "rank", "leading_edge", p=0, note=" (exponent=0)")
report("enrichit_multilevel_p2", "ID", "enrichmentScore", "setSize", "core_enrichment", "rank", "leading_edge", p=2, note=" (exponent=2)")
df_ = report("dose_fgsea", "ID", "enrichmentScore", "setSize", "core_enrichment", "rank", "leading_edge", note=" (legacy DOSE by='fgsea')")
dp_ = report("dose_perm", "ID", "enrichmentScore", "setSize", "core_enrichment", "rank", "leading_edge", note=" (legacy DOSE by='DOSE')")
cp = report("clusterProfiler_GSEA", "ID", "enrichmentScore", "setSize", "core_enrichment", "rank", "leading_edge", note=" (clusterProfiler::GSEA devel)")

# scoreType = "pos": ES must be max of the running sum, leading edge from the top
for name, idc, esc, lec in [("fgsea_multilevel_pos", "pathway", "ES", "leadingEdge"), ("enrichit_multilevel_pos", "ID", "enrichmentScore", "core_enrichment")]:
    df = load(name)
    if df is None: continue
    df = df.set_index(idc); ok = []; dmax = []
    for sid in df.index:
        members = set(sets[sid]) & set(genes); hits = np.isin(genes, list(members))
        from ref_gsea import running_sum
        rs = running_sum(sv, hits); imax = int(np.argmax(rs))
        dmax.append(abs(df.loc[sid, esc] - rs[imax]))
        le = set(genes[(np.arange(len(genes)) <= imax) & hits])
        ok.append(set(str(df.loc[sid, lec]).split("/")) == le)
    print(f"\n== {name}: max |ES - max(running sum)| = {max(dmax):.2e}; leading edge from top == reference: {sum(ok)}/{len(ok)}")

# cross-engine: NES and p-values (Monte-Carlo, so agreement within noise), sign conventions
print("\n== cross-engine NES / p-value on the two planted sets and 3 random sets")
rows = []
for sid in ["UP1", "DOWN1", "R01", "R02", "R03"]:
    row = {"set": sid}
    for name, df, pc, nc in [("fgsea_ml", fm, "pval", "NES"), ("fgsea_simple", fs, "pval", "NES"), ("enrichit_ml", em, "pvalue", "NES"),
                              ("enrichit_sample", es, "pvalue", "NES"), ("enrichit_permute", ep, "pvalue", "NES"), ("DOSE_fgsea", df_, "pvalue", "NES"), ("DOSE_perm", dp_, "pvalue", "NES")]:
        if df is not None and sid in df.index:
            row[name + " p"] = f"{df.loc[sid, pc]:.3g}"; row[name + " NES"] = f"{df.loc[sid, nc]:.3f}"
    rows.append(row)
print(pd.DataFrame(rows).set_index("set").T.to_string())

# ties
t = json.load(open(os.path.join(out, "ties.json")))
st = pd.Series({k: v[0] if isinstance(v, list) else v for k, v in t["stats"].items()})
st = st.sort_values(ascending=False, kind="stable")     # input-order for ties
hits = np.isin(np.array(st.index), t["set"])
r = gsea_es(st.values, hits)
f_es = t['fgsea_ES'][0] if isinstance(t['fgsea_ES'], list) else t['fgsea_ES']
e_es = t['enrichit_ES'][0] if isinstance(t['enrichit_ES'], list) else t['enrichit_ES']
print(f"\n== ties: reference ES with tied block in input order = {r['ES']:.6f}; fgsea {f_es:.6f}; enrichit {e_es:.6f}")
